corehq/apps/locations/bulk_management.py

```diff
@@ -809,12 +809,21 @@ def update_children(parent_stub):
 
         for child_stub in child_stubs:
             child = child_stub.db_object
-            child.parent = parent_location
             if child_stub.do_delete:
                 # keep track of to be deleted items to delete them in top-to-bottom order
                 to_be_deleted.append(child)
             elif child_stub.needs_save:
                 child.location_type = types_by_code.get(child_stub.location_type)
+                if parent_location:
+                    # refetch parent_location object so that mptt related fields are updated consistently,
+                    #   since we are saving top to bottom, parent_location would not have any pending
+                    #   saves, so this is the right point to refetch the object.
+                    child.parent = SQLLocation.objects.get(
+                        domain=parent_location.domain,
+                        site_code=parent_code
+                    )
```

Comment:
This has the potential to complicate things significantly, but it may be worth caching these lookups, since that'll be a lot of db hits.  Another approach could be to save these breadth-first, that is, level-by-level, then do a bulk get for locations at each level before moving on to the next level.